### ISSUE TITLE
when_all() should take rvalue range, instead of apair of iterators

### ISSUE THREAD

**Issue by nyh** [scylladb/seastar#622]
We have overloads of `when_all()` (and `when_all_succeed()`) taking a begin and end iterator, usually belonging to a vector of futures.

However, this API makes two things unclear to users:
1. If these iterators point to some collection or generator of futures (e.g., iterators of an `std::vector<seastar::future>`, how long does object need to live? Is the iterator iterated immediately before waiting, or does it need to live until the future returned by `when_all()` resolves?
2. If these iterators belong to a std::vector<seastar::future>, when `when_all()` is called, what is the state of this vector? Are the contents of this vector still valid? What happens? The short answer is that the user must not try to use these futures again (a future cannot be used once) but this isn't made clear by the API.

A better API which will clarify both questions will be to pass an **rvalue** (i.e., &&) **range** (i.e., something which has begin() and end()) to the API instead of a pair of iterators. The user will need to *move* a vector of futures into when_all(), and it will be clear that he no longer needs to maintains this vector's lifetime, and moreover, he will no longer have access to the futures in that vector and wrongly use them twice.

**Comment by pdziepak** [comment]
s/he/they/

I am not sure accepting a range by rref really clarifies what is going to happen. What if it is a view? E.g.:
```c++
auto integers = { 1, 2, 3, 4 };
return when_all(integers
                | view::transform(int_to_second)
                | view::transform([] (auto x) { return sleep(x); }));
```
It is as unclear whether this is legal as it was with the iterators. The point is that it is not possible to express everything in the function signature and sometimes we need some additional information in the comments.

If a function accepts iterators there is nothing wrong with it modifying the range (e.g. `std::sort`, `std::move`, `std::remove_if`). The caller needs to take care as to what gets invalidated. The iterator overload makes it easier to run when_all() on only part of the range, if the users wishes to do that. This is harder to do with range overload so there is some value in the former.

Forcing the user to move the vector also means that they cannot reuse the storage, this can be alleviated by creating a view of the range to "cheat" `when_all()` but certainly isn't pretty.

All in all, I don't think there is any flaw in the iterator overload that the range one would solve. We probably should have range overload for convenience but I think it should take an universal reference. The everything else can be clarified in the comments.

A separate issue is that like the compile-time sized `when_all()` range-based `when_all()` is very hard to use correctly in case exceptions may be thrown.

**Comment by nyh** [comment]
I agree with your first example, that just an "rvalue range" doesn't guarantee safefty if the range itself keeps pointers to other objects instead of owning it's own storage as vector does :-( It would have helped in the more usual case where one creates a vector<future> just for collecting futures in a loop, and then passes this vector to when_all(), never intending to reuse this vector. Would have been more convenient, as well as clearer what is actually guaranteed, to allow moving the vector into when_all().

I guess that adding a comment guaranteeing that when_all() iterates on the  iterator and moves out all the futures *before* returning (and making sure this is indeed the case!) is the only safe approach...
